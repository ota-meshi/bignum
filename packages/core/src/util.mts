/** compare function */
export function compare(a: bigint, b: bigint): 0 | 1 | -1 {
  return a === b ? 0 : a > b ? 1 : -1;
}

/** Get length */
export function length(a: bigint): number {
  return String(abs(a)).length;
}

/** Get max value */
export function max(a: bigint, b: bigint): bigint {
  return a >= b ? a : b;
}

/** Get abs value */
export function abs(a: bigint): bigint {
  return a >= 0n ? a : -a;
}

/** Checks whether the given value is an even */
export function isEven(a: bigint): boolean {
  return !(a & 1n);
}

/** Find the greatest common divisor. */
export function gcd(a: bigint, b: bigint): bigint {
  a = abs(a);
  b = abs(b);
  while (b) {
    a %= b;
    if (!a) return b;
    b %= a;
  }
  return a;
}
