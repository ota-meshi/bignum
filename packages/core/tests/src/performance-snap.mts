import chai from "chai";
import { jestSnapshotPlugin } from "mocha-chai-jest-snapshot";
import { BigNum } from "../../src/index.mjs";
import { length } from "../../src/util.mts";

if (process.argv.includes("--update")) {
  chai.use(jestSnapshotPlugin());

  describe("performance tests", () => {
    for (const t of [
      () => BigNum.valueOf(100).pow(123.456),
      () => {
        // setup
        let a: BigNum;
        return [
          () => (a = BigNum.valueOf(20.56).pow(48.723)),
          () => a.toString(),
        ];
      },
      () => {
        // setup
        const n = 123456789n ** 1234n;
        return () => {
          for (let i = 0; i < 100; i++) {
            length(n);
          }
        };
      },
    ]) {
      const tests = [{ name: String(t), t, n: 1 }];
      while (tests.length) {
        const test = tests.shift()!;

        if (!String(test.t).includes("setup")) {
          it(test.name, () => {
            let time;
            try {
              const start = performance.now();
              test.t();
              const end = performance.now();
              time = end - start;
            } catch (e) {
              chai.expect(e).toMatchSnapshot();
              return;
            }
            chai.expect(time).toMatchSnapshot();
          });
          continue;
        }
        tests.unshift(
          ...([test.t()].flat() as (() => any)[]).map((child) => {
            return {
              name: `${test.name}->\n${"  ".repeat(test.n)}${String(child)
                .split("\n")
                .join(`\n${"  ".repeat(test.n)}`)}`,
              t: child,
              n: test.n + 1,
            };
          }),
        );
      }
    }
  });
}
