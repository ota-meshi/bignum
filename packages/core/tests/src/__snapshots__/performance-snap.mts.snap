// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`performance tests () => {
                // setup
                const a = BigNum.valueOf(20.56);
                return () => a.pow(48.723);
            }
  ->
  () => a.pow(48.723) 1`] = `566.5785000000001`;

exports[`performance tests () => {
                // setup
                const a = BigNum.valueOf(123456789.12345678);
                return () => {
                    for (let i = 0; i < 100; i++) {
                        a.sqrt();
                    }
                };
            }
  ->
  () => {
                      for (let i = 0; i < 100; i++) {
                          a.sqrt();
                      }
                  } 1`] = `2.021041999999852`;

exports[`performance tests () => {
                // setup
                const n = 123456789n ** 1234n;
                return () => {
                    for (let i = 0; i < 100; i++) {
                        length(n);
                    }
                };
            }
  ->
  () => {
                      for (let i = 0; i < 100; i++) {
                          length(n);
                      }
                  } 1`] = `28.086583999999903`;

exports[`performance tests () => {
                // setup
                const n = 123456789n ** 1234n;
                return () => {
                    for (let i = 0; i < 100; i++) {
                        n.toString();
                    }
                };
            }
  ->
  () => {
                      for (let i = 0; i < 100; i++) {
                          n.toString();
                      }
                  } 1`] = `28.040959000000157`;

exports[`performance tests () => {
                // setup
                let a;
                return [
                    () => (a = BigNum.valueOf(20.56).pow(48.723)),
                    () => a.toString(),
                ];
            }
  ->(1)
  () => (a = BigNum.valueOf(20.56).pow(48.723)) 1`] = `518.5813330000001`;

exports[`performance tests () => {
                // setup
                let a;
                return [
                    () => (a = BigNum.valueOf(20.56).pow(48.723)),
                    () => a.toString(),
                ];
            }
  ->(2)
  () => a.toString() 1`] = `155.21400000000017`;

exports[`performance tests () => BigNum.valueOf(100).pow(123.456) 1`] = `2.5096669999998085`;
